### PR TITLE
tree_hash: more efficient file hashing

### PR DESCRIPTION
**Pull request by StefanKarpinski**
This avoids reading the entire contents of a file into memory just
to hash it, instead hashing the data incrementally, using the same
logic as read_data. A future improvement would be to actually use
read_data by wrapping SHA hashing in an I/O-like interface.

### CODE REVIEW

File: src/extract.jl [hunk 1]
```diff
@@ -145,6 +143,28 @@ function git_object_hash(
...
+    buf::Vector{UInt8} = Vector{UInt8}(undef, DEFAULT_BUFFER_SIZE),
+)
+    ctx = HashType()
+    SHA.update!(ctx, codeunits("blob $size\0"))
+    # TODO: this largely duplicates the logic of read_data
+    # read_data could be used directly if SHA offered an interface
+    # where you write data to an IO object and it maintains a hash
+    while size > 0
+        r = readbytes!(tar, buf, size < sizeof(buf) ? round_up(size) : sizeof(buf))
+        r < 512 && eof(io) && error("premature end of tar file")
```
Comment (GunnarFarneback) [inline src/extract.jl:159]: This 512 doesn't make a whole lot of sense. Missed when making the `buf` size configurable?

Comment (StefanKarpinski) [reply src/extract.jl:159]: @keno: was this missed or is it intentional? Note: copied from your `read_data` method.

Comment (StefanKarpinski) [reply src/extract.jl:159]: I guess the more general version here seems like we should compute how much we're looking to read and if we're able to read less than that it's a premature end of file. Testing error conditions is such a pain, I really don't want to...

Comment (Keno) [reply src/extract.jl:159]: Tar is written 512 byte sectors, so if a tar file isn't a multiple of 512 it's corrupt. Should probably check the size mod 512 though

Comment (StefanKarpinski) [reply src/extract.jl:159]: Potential fix: https://github.com/JuliaIO/Tar.jl/pull/40

**Pull request merged**